python/ray/util/client/dataclient.py

```diff
@@ -60,48 +60,102 @@ def _start_datathread(self) -> threading.Thread:
+        elif response.req_id in self.asyncio_waiting_data:
+            # Async response. Check for callback and update outstanding
+            # requests.
+            callback = self.asyncio_waiting_data.pop(response.req_id)
+            try:
+                if callback:
+                    callback(response)
+            except Exception:
+                logger.exception("Callback error:")
+            if response.req_id in self.outstanding_requests:
+                del self.outstanding_requests[response.req_id]
+        else:
+            # Blocking response. Populate ready data and wake up waiting
+            # threads.
             with self.cv:
-                self._in_shutdown = True
+                self.ready_data[response.req_id] = response
                 self.cv.notify_all()
-            if e.code() == grpc.StatusCode.CANCELLED:
-                # Gracefully shutting down
-                logger.info("Cancelling data channel")
-            elif e.code() in (grpc.StatusCode.UNAVAILABLE,
-                              grpc.StatusCode.RESOURCE_EXHAUSTED):
-                # TODO(barakmich): The server may have
-                # dropped. In theory, we can retry, as per
-                # https://grpc.github.io/grpc/core/md_doc_statuscodes.html but
-                # in practice we may need to think about the correct semantics
-                # here.
-                logger.info("Server disconnected from data channel")
-            else:
-                logger.exception(
-                    "Got Error from data channel -- shutting down:")
+
+    def _process_rpc_error(self, e: grpc.RpcError) -> bool:
+        """
+        Processes RPC errors that occur while reading from data stream.
+        Returns True if the error can be recovered from, False otherwise.
+        """
+        if not self.client_worker._can_reconnect(e):
+            logger.info("Unrecoverable error in data channel.")
+            return False
+
+        # Ping the server to see if the current channel is reuseable
+        try:
+            ping_succeeded = self.client_worker.ping_server(timeout=5)
+        except grpc.RpcError:
+            ping_succeeded = False
+
+        if not ping_succeeded:
+            # Ping failed, try refreshing the data channel
+            logger.warning(
+                "Encountered connection issues in the data channel. "
+                "Attempting to reconnect.")
+            logger.debug(e)
+            try:
+                self.client_worker._connect_channel(reconnecting=True)
+            except ConnectionError:
+                logger.warning("Failed to reconnect the data channel")
+                return False
+            logger.info("Reconnection succeeded!")
+
+        # Recreate the request queue, and resend outstanding requests
+        self.request_queue = queue.Queue()
+        with self.outstanding_requests_lock:
+            for request in self.outstanding_requests.values():
```

Comment:
This can reorder outstanding requests, which becomes problematic when we schedule asynchronously. I wonder if we can build an implementation of the queue interface that iterates in order and only deletes when a response is received. That way we don't need to store outstanding_requests separately.

Reply:
Worked around by making outstanding_requests an ordered queue, and inserting into outstanding_requests while a lock is held when generating request_ids/inserting into the queue.